### PR TITLE
Improve import error message for stubgen Issue 6794

### LINKED ISSUE #6794: stubgen -m/-p "Failed to import xx skipping it" because of ModuleNotFoundError
`stubgen -m my_module`
`stubgen -p my_package `
output: "Failed to import my_module/my_package skipping it"  Error without details.
Because the environment actually missing packages which generate ModuleNotFoundError.

mypy 0.701
python3.7

### PR DESCRIPTION
**Pull request by xbian0403**
Since the error message is being printing out, any good ideas on how to test it?

`pytest -n0 --mypy-verbose -k testImportErrorMessage_import -s`
currently error message looks like this:
Failed to import a with error: ("No module named 'b'",); skipping it

Fixes #6794

### CODE REVIEW

File: mypy/test/teststubgen.py [hunk 1]
```diff
@@ -266,6 +266,7 @@ def run_case(self, testcase: DataDrivenTestCase) -> None:
             with open(file, 'w') as f:
                 f.write(content)
 
+        sys.path.append('.')
```
Comment (ilevkivskyi) [inline mypy/test/teststubgen.py]: Why is this change needed? Is it here so that tests mimic the logic of `mypy.stubgen.main()`? Please add a comment.

File: test-data/unit/stubgen.test [hunk 2]
```diff
@@ -1435,3 +1435,12 @@ from typing import Any
...
     def other(self) -> Any: ...
+
+[case testImportErrorMessage_import]
+# flags: --module a
+
+[file a.py]
+import b
+
+
+[out]
```
Comment (ilevkivskyi) [inline test-data/unit/stubgen.test]: Missing newline at the end of file.

File: test-data/unit/stubgen.test [hunk 3]
```diff
@@ -1435,3 +1435,12 @@ from typing import Any
...
     @abstractmethod
     def other(self) -> Any: ...
+
+[case testImportErrorMessage_import]
+# flags: --module a
+
+[file a.py]
+import b
+
+
```
Comment (ilevkivskyi) [inline test-data/unit/stubgen.test]: These empty lines are not needed.

File: test-data/unit/stubgen.test [hunk 4]
```diff
@@ -1435,3 +1435,12 @@ from typing import Any
 class C(Base, metaclass=abc.ABCMeta):
     @abstractmethod
     def other(self) -> Any: ...
+
+[case testImportErrorMessage_import]
```
Comment (ilevkivskyi) [inline test-data/unit/stubgen.test]: What exactly this test tests? Does it fail for the current version? (I assume it doesn't). Ideally I would add a test that checks the error formatting (or at least something that fails on master, but passes on your branch).

File: mypy/stubutil.py [hunk 5]
```diff
@@ -155,8 +157,11 @@ def generate_guarded(mod: str, target: str,
             print('Created %s' % target)
 
 
-def report_missing(mod: str) -> None:
-    print('Failed to import {}; skipping it'.format(mod))
+def report_missing(mod: str, message: Optional[str] = None) -> None:
+    if message:
+        print('Failed to import {} with error: {}; skipping it'.format(mod, message))
+    else:
+        print('Failed to import {}; skipping it'.format(mod))
```
Comment (ilevkivskyi) [inline mypy/stubutil.py]: I would avoid the duplication by making the default value for `message` an empty string (in the function signature) and writing this as:
```python
if message:
    message = ' with error: ' + message
print('Failed to import {}{}; skipping it'.format(mod, message))
```

File: mypy/test/teststubgen.py (hunk 1 above)
Comment (xbian0403) [reply mypy/test/teststubgen.py]: current folder is not in sys path, when I test the function generate_stubs with `--module` option which should call `importlib.import_module` can't find the module. Thus I add it to sys.path.

File: test-data/unit/stubgen.test (hunk 4 above)
Comment (xbian0403) [reply test-data/unit/stubgen.test]: Actually I need help on testing it. The error message is being print out. Any suggestion for testing it?
should I capture stdout and check it for only this test?
`
          capturedOutput = io.StringIO()
          sys.stdout = capturedOutput
`

Comment (ilevkivskyi) [reply test-data/unit/stubgen.test]: > should I capture stdout and check it for only this test?

I think this is OK, just don't forget to put the old value back in a `try ... finally: ...`. You can make this testing feature more generic for stubgen tests. For example you can run each test with captured output and add an additional check that stdout matches some regexp if the test case contains a `[stdout <some regexp here>]` line.

Alternatively you can simply make it a non data-driven test case similar to `mypy.teststubgen.StubgenCmdLineSuite.test_packages_found()` (this is probably even better).

File: mypy/test/teststubgen.py [hunk 6]
```diff
@@ -64,6 +64,23 @@ def test_packages_found(self) -> None:
             finally:
                 os.chdir(current)
 
+    def test_module_not_found(self) -> None:
+        current = os.getcwd()
+        with tempfile.TemporaryDirectory() as tmp:
+            try:
+                import io
+                capturedOutput = io.StringIO()
```
Comment (ilevkivskyi) [inline mypy/test/teststubgen.py]: This is a non-typical case for Python variables, I would propose to use `captured_output`.

File: mypy/test/teststubgen.py [hunk 7]
```diff
@@ -64,6 +64,23 @@ def test_packages_found(self) -> None:
             finally:
                 os.chdir(current)
 
+    def test_module_not_found(self) -> None:
+        current = os.getcwd()
+        with tempfile.TemporaryDirectory() as tmp:
+            try:
+                import io
```
Comment (ilevkivskyi) [inline mypy/test/teststubgen.py]: Please put this import at the top of file.

File: mypy/test/teststubgen.py [hunk 8]
```diff
@@ -64,6 +64,23 @@ def test_packages_found(self) -> None:
...
+                capturedOutput = io.StringIO()
+                sys.stdout = capturedOutput
```
Comment (ilevkivskyi) [inline mypy/test/teststubgen.py]: I would move the parts that should not fail (the two above lines) above the `try`.

File: mypy/test/teststubgen.py [hunk 9]
```diff
@@ -64,6 +64,23 @@ def test_packages_found(self) -> None:
...
+                os.chdir(tmp)
+                self.make_file(tmp, 'mymodule.py', content='import a')
+                opts = parse_options(['-m', 'mymodule'])
+                py_mods, c_mods = collect_build_targets(opts, mypy_options(opts))
+                sys.stdout = sys.__stdout__
```
Comment (ilevkivskyi) [inline mypy/test/teststubgen.py]: Please put this line in the `finally` block, since this is a clean-up step that should be always taken.

File: mypy/test/teststubgen.py [hunk 10]
```diff
@@ -64,6 +64,23 @@ def test_packages_found(self) -> None:
...
+                match = re.search('No module named \'a\'', capturedOutput.getvalue())
```
Comment (ilevkivskyi) [inline mypy/test/teststubgen.py]: Use double quotes `"` for outer position to avoid `\`.

File: mypy/test/teststubgen.py [hunk 11]
```diff
@@ -64,6 +64,23 @@ def test_packages_found(self) -> None:
...
+                self.assertTrue(match)
```
Comment (ilevkivskyi) [inline mypy/test/teststubgen.py]: You can probably simply use `self.assertRegex()` instead of the two above lines.

**Comment by gvanrossum** [comment]
Thanks for the fix. Very useful!
